akka-cluster-tools/src/main/scala/akka/cluster/singleton/ClusterSingletonManager.scala

```diff
@@ -683,7 +731,80 @@ class ClusterSingletonManager(
       self ! DelayedMemberRemoved(m)
   }
 
-  def gotoOldest(): State = {
+  // Try and go to oldest, taking the lease if needed
+  def tryGoToOldest(): State = {
+    // check if lease
+    lease match {
+      case None ⇒
+        goToOldest()
+      case Some(_) ⇒
+        logInfo("Trying to acquire lease before starting singleton")
+        tryAcquireLease()
+    }
+  }
+
+  when(AcquiringLease) {
+    case Event(AcquireLeaseResult(result), _) ⇒
+      logInfo("Acquire lease result {}", result)
+      if (result) {
+        goToOldest()
+      } else {
+        setTimer(LeaseRetryTimer, LeaseRetry, leaseRetryInterval)
+        stay using AcquiringLeaseData(leaseRequestInProgress = false, None)
+      }
+    case Event(Terminated(ref), AcquiringLeaseData(_, Some(singleton))) if ref == singleton ⇒
+      logInfo("Singleton actor terminated. Trying to acquire lease again before re-creating.")
+      tryAcquireLease()
+    case Event(AcquireLeaseFailure(t), _) ⇒
+      log.error(t, "failed to get lease (will be retried)")
+      setTimer(LeaseRetryTimer, LeaseRetry, leaseRetryInterval)
+      stay using AcquiringLeaseData(leaseRequestInProgress = false, None)
+    case Event(LeaseRetry, _) ⇒
+      tryAcquireLease()
```

Comment:
couldn't `Some(singleton)` from AcquiringLeaseData be lost here since tryAcquireLease is using None

Reply:
We don't try and acquire the lease until the old instance has Terminated, and we only schedule a retry when the acquire fails.

Maybe this would be clearer with another state `AwaitingTerminationBeforeAcquire`? but there'd be a lot of common cases

Reply:
Ok, add some comments

Reply:
:+1: